src/part-guide/adv-async-await.md

```diff
@@ -2,26 +2,142 @@
+
+There are many other ways to block a whole thread, and we'll come back to blocking several times in this guide.
+
+### Cancellation
+
+Cancellation means stopping a future (or task) from executing. Since in Rust, futures must be driven forward by an external force (like the async runtime), if a future is no longer driven forward then it will not execute any more. If a future is dropped (remember, a future is just a plain old Rust object), then it can never make any more progress and is cancelled.
+
+Cancellation can be initiated in a few ways:
+
+- calling [`abort`](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.abort) on a task's 'JoinHandle' (or an `AbortHandle`),
+- via a [`CancellationToken`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html) (which requires the future being cancelled to notice the token and cooperatively cancel itself),
+- implicitly, by a function or macro like [`select`](https://docs.rs/tokio/latest/tokio/macro.select.html),
+- by simply dropping a future if you have a direct reference to it.
+
+The first two are specific to Tokio, though most runtimes provide similar facilities. The second requires cooperation of the future being cancelled, but the others do not. In these other cases, the cancelled future will get no notification of cancellation and no opportunity to clean up (besides its destructor). Note that even if a future has a cancellation token, it can still be cancelled via the other methods which won't trigger the cancellation token.
+
+From the perspective of writing async code (in async functions, blocks, futures, etc.), the code might stop executing at any `await` (including hidden ones in macros) and never start again. In order for your code to be correct (specifically to be *cancellation safe*), it must never leave any data in an inconsistent state at any await point.
+
+An example of how this can go wrong is if an async function reads data into an internal buffer, then awaits the next datum. If reading the data is destructive (i.e., cannot be re-read from the original source) and the async function is cancelled, then the internal buffer will be dropped, and the data in it will be lost.
+
+We'll be coming back to cancellation and cancellation safety a few times in this guide, and there is a whole [chapter]() on the topic in the reference section.
+
+
+## Async blocks
+
+A regular block in Rust (`{ ... }`) groups code together in the source and creates a scope of encapsulation for names. At runtime, the block is executed in order and evaluates to the value of it's last expression (or void (`()`) if there is no trailing expression).
+
+Similarly to async functions, an async block is a deferred version of a regular block. An async block scopes code and names together, but at runtime it is not immediately executed and evaluates to a future. To execute the block and obtain the result, it must be `await`ed. E.g.,
```

Comment:
I would expect a comma here as that's the usual usage for `e.g.,` and we're not followed by a list, but by a figure

Reply:
We use a colon for this elsewhere, e.g. in the Reference.  We'd say, "For example:" rather than "For example," when followed by a code block.  Or, "For example, consider this case:".

With the comma there, I'd expect the sentence to continue after the figure, or I'd expect the figure to end with a period (as is sometimes seen in mathematical writing).

The MLA seems [to follow](https://style.mla.org/colons-how-to-use-them/) this style, e.g.:

> Use colons to introduce a quotation when it is not integrated into the syntax of your sentence or otherwise requires a formal introduction:
>
> ...
>
> Use a colon to introduce the formal expression of a rule or principle:

Maybe more importantly, this is what Knuth [suggests](https://jmlr.csail.mit.edu/reviewing-papers/knuth_mathematical_writing.pdf) also (page 3):

> The statement just preceding a theorem, algorithm, etc., should be a complete sentence or should end with a colon.